File: tests/money.test.ts
import { describe, expect, it } from "vitest";

import { detectGrade, formatMoney, parseMoney } from "@/lib/money";

describe("parseMoney", () => {
  it("reads an eBay US price cell", () => {
    expect(parseMoney("US $124.99")).toEqual({ amount: 124.99, currency: "USD" });
  });

  it("strips thousands separators", () => {
    expect(parseMoney("$1,250.00")).toEqual({ amount: 1250, currency: "USD" });
  });

  it("reads a pound price", () => {
    expect(parseMoney("£89.50")).toEqual({ amount: 89.5, currency: "GBP" });
  });

  it("treats a trailing comma-two-digits as a decimal separator", () => {
    expect(parseMoney("43,50 €")).toEqual({ amount: 43.5, currency: "EUR" });
  });

  it("handles European grouping and decimals together", () => {
    expect(parseMoney("€1.250,00")).toEqual({ amount: 1250, currency: "EUR" });
  });

  it("recognises a bare currency code", () => {
    expect(parseMoney("EUR 43.50")).toEqual({ amount: 43.5, currency: "EUR" });
  });

  it("returns null when there is no number", () => {
    expect(parseMoney("Best offer accepted")).toBeNull();
  });

  it("returns null for empty input", () => {
    expect(parseMoney("")).toBeNull();
  });
});

describe("formatMoney", () => {
  it("formats a USD amount", () => {
    expect(formatMoney(124.5, "USD")).toBe("$124.50");
  });

  it("renders a dash for a missing amount", () => {
    expect(formatMoney(null)).toBe("—");
  });
});

describe("detectGrade", () => {
  it("finds a PSA grade", () => {
    expect(detectGrade("Charizard ex 054/165 PSA 10 GEM MINT")).toBe("PSA 10");
  });

  it("normalises spacing and casing", () => {
    expect(detectGrade("charizard psa10 mint")).toBe("PSA 10");
  });

  it("reads half grades", () => {
    expect(detectGrade("Pikachu BGS 9.5")).toBe("BGS 9.5");
  });

  it("recognises an explicitly ungraded listing", () => {
    expect(detectGrade("Charizard 4/102 raw near mint")).toBe("Ungraded");
  });

  it("returns null when no grade is mentioned", () => {
    expect(detectGrade("Charizard 4/102 Base Set")).toBeNull();
  });
});

File: android/core/src/test/kotlin/com/paladex/ex/core/MoneyTest.kt
package com.paladex.ex.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class MoneyTest {

    @Test
    fun `reads an eBay US price cell`() {
        assertEquals(Money.Parsed(124.99, Currency.USD), Money.parse("US \$124.99"))
    }

    @Test
    fun `strips thousands separators`() {
        assertEquals(Money.Parsed(1250.0, Currency.USD), Money.parse("\$1,250.00"))
    }

    @Test
    fun `reads a pound price`() {
        assertEquals(Money.Parsed(89.5, Currency.GBP), Money.parse("£89.50"))
    }

    @Test
    fun `treats a trailing comma-two-digits as a decimal separator`() {
        assertEquals(Money.Parsed(43.5, Currency.EUR), Money.parse("43,50 €"))
    }

    @Test
    fun `handles European grouping and decimals together`() {
        assertEquals(Money.Parsed(1250.0, Currency.EUR), Money.parse("€1.250,00"))
    }

    @Test
    fun `recognises a bare currency code`() {
        assertEquals(Money.Parsed(43.5, Currency.EUR), Money.parse("EUR 43.50"))
    }

    @Test
    fun `returns null when there is no number`() {
        assertNull(Money.parse("Best offer accepted"))
    }

    @Test
    fun `returns null for blank input`() {
        assertNull(Money.parse(""))
        assertNull(Money.parse(null))
    }

    @Test
    fun `takes the low end of a price range`() {
        assertEquals(10.0, Money.parsePriceCell("\$10.00 to \$25.00")?.amount)
    }

    @Test
    fun `formats a USD amount`() {
        assertEquals("\$124.50", Money.format(124.5, Currency.USD))
    }

    @Test
    fun `renders a dash for a missing amount`() {
        assertEquals("—", Money.format(null))
    }

    @Test
    fun `finds a PSA grade`() {
        assertEquals("PSA 10", Money.detectGrade("Charizard ex 054/165 PSA 10 GEM MINT"))
    }

    @Test
    fun `normalises grade spacing and casing`() {
        assertEquals("PSA 10", Money.detectGrade("charizard psa10 mint"))
    }

    @Test
    fun `reads half grades`() {
        assertEquals("BGS 9.5", Money.detectGrade("Pikachu BGS 9.5"))
    }

    @Test
    fun `recognises an explicitly ungraded listing`() {
        assertEquals("Ungraded", Money.detectGrade("Charizard 4/102 raw near mint"))
    }

    @Test
    fun `returns null when no grade is mentioned`() {
        assertNull(Money.detectGrade("Charizard 4/102 Base Set"))
    }
}
